### DISCUSSION TITLE
Message queue (pub/sub) and NextJS

### DISCUSSION THREAD

**Discussion by mikaa123** [vercel/next.js#14603] [Help]
Hey everyone,

I'm wondering how you'd include a message queue without your NextJS application. For example, I'd like send email notifications once an action happened. The email notifications can be sent in the background, it doesn't need to be blocking. At the moment, I'm simply calling a lambda (/api/notify) and don't wait for the response. I'd be curious to hear best practices.

Thanks!

**Comment by nalanj** [comment]
I noticed this post while looking for similar. For me it came down to wanting to take advantage of the build tooling Next.js provides, so I just added an extra entry in my webpack config and then run it in its own container with a pretty standard queue setup.

**Comment by shamoons** [comment]
> I noticed this post while looking for similar. For me it came down to wanting to take advantage of the build tooling Next.js provides, so I just added an extra entry in my webpack config and then run it in its own container with a pretty standard queue setup.

Can you share how you did this?

**Comment by nalanj** [comment]
It's a little convoluted, but here's broadly what I added to my `next.config.js`:

```js
// extra file entries for server-side builds
const serverEntries = {
  queue: ["./src/queue.js"],
  cron: ["./src/cron.js"],
};

...


module.exports = {
   ...
   webpack(config, { isServer, nextRuntime }) {
     ...
     return {
        ...config,
        entry() {
          return config.entry().then((entry) => {
            return Object.assign(
              {}, entry, isServer && nextRuntime === "nodejs" ? serverEntries : {}
            );
          });
        }
     };
   })
   ...
};
```

**Comment by deadcoder0904** [comment]
@nalanj is this the best solution? is it reliable?

i was looking at instrumentation but it's still experimental.

@mikaa123 @shamoons did you find an alternative?

**Reply by nalanj** [reply]
@deadcoder0904 I've been using this for a couple of years now. It's about the best I've been able to figure out.

**Reply by deadcoder0904** [reply]
@nalanj does this queue run all the time? as long as next.js is alive? i want to use like bullmq + docker + next.js to add an email to a queue after payment is done but don't want to send the email directly as its a non-critical task.

is this a better idea for my stack? or a separate node.js server is? i mean i'm anyways using dockerfile.

**Reply by nalanj** [reply]
It doesn't run it at all, it simply adds it to the webpack build entrypoints so that you end up with a script that you can run yourself however you'd like.
